realm/realm-library/src/main/cpp/io_realm_internal_OsObject.cpp

```diff
@@ -204,3 +291,78 @@ JNIEXPORT void JNICALL Java_io_realm_internal_OsObject_nativeStopListening(JNIEn
     }
     CATCH_STD()
 }
+
+JNIEXPORT jlong JNICALL Java_io_realm_internal_OsObject_nativeCreateRow(JNIEnv* env, jclass, jlong shared_realm_ptr,
+                                                                        jlong table_ptr)
+{
+    try {
+        return do_create_row(shared_realm_ptr, table_ptr);
+    }
+    CATCH_STD()
+    return -1;
+}
+
+JNIEXPORT jlong JNICALL Java_io_realm_internal_OsObject_nativeCreateNewObject(JNIEnv* env, jclass,
+                                                                              jlong shared_realm_ptr, jlong table_ptr)
+{
+    try {
+        size_t row_ndx = do_create_row(shared_realm_ptr, table_ptr);
+        auto& table = *(reinterpret_cast<realm::Table*>(table_ptr));
+        return reinterpret_cast<jlong>(new Row(table[row_ndx]));
+    }
+    CATCH_STD()
+    return 0;
+}
+
+JNIEXPORT jlong JNICALL Java_io_realm_internal_OsObject_nativeCreateNewObjectWithPrimaryKey__JJJJZ(
+    JNIEnv* env, jclass, jlong shared_realm_ptr, jlong table_ptr, jlong pk_column_ndx, jlong pk_value,
+    jboolean is_pk_null)
+{
+    try {
+        auto& table = *(reinterpret_cast<realm::Table*>(table_ptr));
+        size_t row_ndx =
+            do_create_row_with_primary_key(env, shared_realm_ptr, table_ptr, pk_column_ndx, pk_value, is_pk_null);
+        if (row_ndx != npos) {
+            return reinterpret_cast<jlong>(new Row(table[row_ndx]));
+        }
+    }
+    CATCH_STD()
+    return 0;
+}
+
+JNIEXPORT jlong JNICALL Java_io_realm_internal_OsObject_nativeCreateRowWithPrimaryKey__JJJJZ(
+    JNIEnv* env, jclass, jlong shared_realm_ptr, jlong table_ptr, jlong pk_column_ndx, jlong pk_value,
+    jboolean is_pk_null)
+{
+    try {
+        return do_create_row_with_primary_key(env, shared_realm_ptr, table_ptr, pk_column_ndx, pk_value, is_pk_null);
+    }
+    CATCH_STD()
+    return npos;
```

Comment:
It is actually a value defined in core. See `array.hpp`. Will it be less confused with namespace here? like `return realm::npos`